src/docker.rs

```diff
@@ -0,0 +1,167 @@
+use anyhow::{anyhow, Result};
+use async_trait::async_trait;
+use docker_api::api::{ContainerCreateOpts, PullOpts, RegistryAuth, RmContainerOpts};
+use futures_util::{StreamExt, TryStreamExt};
+use std::path::PathBuf;
+use tar::Archive;
+
+use super::image;
+
+pub struct DockerImage {
+    pub image: String,
+    pub repo: String,
+    pub tag: String,
+    pub runtime: docker_api::Docker,
+}
+
+#[async_trait]
+impl image::Image for DockerImage {
+    // pull ensures that the image is present locally and, if it is isn't
+    // will do the work necessary to pull it.
+    async fn pull(&self, username: String, password: String, force: bool) -> Result<()> {
+        if self.is_present_locally().await && !force {
```

Comment:
IMO this logic is a bit convoluted -- I would prefer splitting up the two clauses into separate statements for clarity

Reply:
Makes sense!